### ISSUE TITLE
FeatureLayerService timeout option

### ISSUE THREAD

**Issue by nathanhilbert** [Esri/esri-leaflet#728]
I'm hitting some issues with 2.0.0-beta.7 of featurelayer requests taking up to 5 seconds for certain areas, then returning a 500 from ArcGIS Online.  This is definitely a problem with ArcGIS Online, but I was thinking a workaround for similar slow featurelayerservices would be useful.  

There are two issues that I have seen:

1. I have a featureservice that has a "where" set.  The user clicks on a button, and it updates the "where" to a new query.  If the user clicks on the different buttons quickly, early responses will pend while new calls are being generated and completed.  When that pending call finishes, it triggers the "load" event and draws the geometry of that pended call.  

2. To avoid users from clicking on buttons while loads are happening, I put a flag to ignore clicks while loading.  However, this still hits the issue that if a feature request is pending for five seconds, the user is sitting there waiting.

My first thought was to put a timeout on each "requeststart" event.  If it takes more than 2 seconds, then just abort it.  I can't quite find how to access the httprequest object to call abort.  Maybe someone can point me in the right direction.  Maybe it would be useful to have an option that does this internally and fires necessary events.  Any thoughts would helpful.  Thanks.

abbreviated code:
```
    featureLayer= L.esri.featureLayer({
              url: 'myAGOfeatureserver',
              where: "1=1",
              style: serviceStyle,
              simplifyFactor: 0.1,
              where: "Name='poly001'",
              cacheLayers: false
              });



    featureLayer.on("loading", function(ev){
      currentlyloading = true;
      $("#loading").show();
    });


    featureLayer.on("load", function(ev){
      currentlyloading = false;
      $("#loading").hide();
    });

    featureLayer.service.on("requeststart", function(ev){
      console.log(ev);
      console.log(ev['target']);
     //where's abort?
    });


    featureLayer.addTo(map);

```

**Comment by jgravois** [comment]
i'm hopeful that when we sort out #719 you'll see a performance improvement, but either way, it sounds like a good idea to me t

> Maybe it would be useful to have an option that does this internally and fires necessary events.

what do you mean by 'option'?  my gut reaction to this is that we should ensure all pending requests are aborted when [`featureLayer.setWhere()`](https://github.com/Esri/esri-leaflet/blob/master/src/Layers/FeatureLayer/FeatureManager.js#L188) is called automatically (for everyone all the time).

**Comment by nathanhilbert** [comment]
Thanks for the response.  
I was thinking more an edit [here](https://github.com/Esri/esri-leaflet/blob/master/src/Request.js#L78) would add 
```
  httpRequest.timeout = optinons.timeout;  //using 1500

  httpRequest.ontimeout = function(){
    this.onerror();
  }
```

From the API, maybe something like this
```
featureLayer = L.esri.featureLayer();
featureLayer.service.setTimeout(1500);

//or

featureLayer.service.timeout = 1500;
```

**Comment by patrickarlt** [comment]
> I have a featureservice that has a "where" set. The user clicks on a button, and it updates the "where" to a new query. If the user clicks on the different buttons quickly, early responses will pend while new calls are being generated and completed.

A long time ago I tried to implement aborting all pending requests when creating new requests either through panning, zooming, `setWhere` or `setTimeRange`. It worked on and aborting JSONP requests was wonky but it also caused most of the tests to inexplicably fail so I abandoned it.

I always wanted to revisit the idea though so maybe it is time to experiment with it again.

>  I can't quite find how to access the httprequest object to call abort.

That is because you can't :smile:. The `XMLHTTPRequest` object ends up getting buried inside feature layer it was a trade off since it is pretty rare you will actually want to access it except in cases like this.

>  Maybe it would be useful to have an option that does this internally and fires necessary events. 

This is definitely the right approach and I see your PR so lets continue the conversation in https://github.com/Esri/esri-leaflet/pull/730.

**Comment by nathanhilbert** [comment]
Closed with #730
